Workflow step definitions

KevinVdV opened the pull request:
PR for the following rest contracts:
* https://github.com/DSpace/Rest7Contract/blob/master/workflowdefinitions.md
* https://github.com/DSpace/Rest7Contract/blob/master/workflowsteps.md
* https://github.com/DSpace/Rest7Contract/blob/master/workflowactions.md

This PR also contains the changes made in: https://github.com/DSpace/DSpace/pull/2420, so this PR should be accepted prior to this one.

MarieVerdonck committed:
67694: new /api/config/workflowsdefintions endpoints:
/api/config/workflowdefinitions/<:definition-name>
/api/config/workflowdefinitions
/api/config/workflowdefinitions/search/findByCollection?uuid=<:collection-uuid>
/api/config/workflowdefinitions/<:definition-name>/collections

Signed-off-by: Marie Verdonck <maria.verdonck@atmire.com>

MarieVerdonck committed:
67694: checkstyle and licenses

MarieVerdonck committed:
67694: Exception handling improvements and /search moved to repository

MarieVerdonck committed:
67694: isDefault fix, workflowdefinitions IT, exception handling

MarieVerdonck committed:
67694: forgot to log error stack itself

MarieVerdonck committed:
- Merged with KevinVdV/DSpace/DS-4239-migrate-workflow-xml-to-spring
- Changed all implementations of XmlWorkflowFactoryImpl to work with new mapping
- Moved their java doc to XmlWorkflowFactory
- Removed/Changed exceptions
- Altered tests in WorkflowDefinitionRestRepositoryIT to work with changes

Merge remote-tracking branch 'github-kevin/DS-4239-migrate-workflow-xml-to-spring' into workflow-step-definitions

# Conflicts:
#	dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowFactoryImpl.java
#	dspace-api/src/main/java/org/dspace/xmlworkflow/factory/XmlWorkflowFactory.java

MarieVerdonck committed:
Merge branch 'w2p-67996-workflow-junit-tests' into workflow-step-definitions

MarieVerdonck committed:
67998: Merged in w2p-67996, altered XmlWorkflowFactoryTest to destroy created comcol

MarieVerdonck committed:
67998: feedback exceptions

MarieVerdonck committed:
67998: tested config handle moved, since other non-collection dso already create at that handle

MarieVerdonck committed:
67998: non-numbered handle config so can't accidentally be created

MarieVerdonck committed:
68150: fix and tests for pagination for api/config/workflowdefinitions

MarieVerdonck committed:
68150: /workflowdefinitions/defaultWorkflow/collections now returns all non-mapped collections and now returns paginated result + tests

MarieVerdonck committed:
68155: /api/config/workflowactions/<:action-name> endpoint + tests

MarieVerdonck committed:
68150: resolve collection handles to collection in XmlWorkflowFactory & forgotten exception

MarieVerdonck committed:
68155: SUBMIT_CANCEL option removed from SelectReviewerAction

MarieVerdonck committed:
68155: xmlWorkflowFactory autowired where possible & WorkflowAction.name > id

MarieVerdonck committed:
68253: GET /api/config/workflowsteps/<:step-name> endpoint

MarieVerdonck committed:
68253: GET /api/config/workflowsteps/<:step-name> IT

MarieVerdonck committed:
68274: /api/config/workflowdefinitions/<:definition-name>/steps endpoint and link

MarieVerdonck committed:
68274: /workflowdefinitions/<:definition-name>/steps endpoint as embedded resource and collections link

KevinVdV committed:
Merge branch 'DS-4239-migrate-workflow-xml-to-spring' into workflow-step-definitions

KevinVdV committed:
Fixing merge issue in the workflow-actions.xml file

benbosman commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/WorkflowDefinitionController.java:
```diff
@@ -0,0 +1,76 @@
+            Context context = ContextUtil.obtainContext(request);
+            List<Collection> collectionsMappedToWorkflow;
+            if (xmlWorkflowFactory.isDefaultWorkflow(workflowName)) {
+                collectionsMappedToWorkflow = xmlWorkflowFactory.getAllNonMappedCollectionsHandles(context);
```
Shouldn't this also return the collection specifically linked to the default workflow (if any)?

benbosman reviewed (commented):
I have reviewed the code (in comparison with https://github.com/DSpace/DSpace/pull/2420)
I have included some missing JavaDocs, and some questions inline

I haven't tested the functionality yet

tdonohue commented:
As #2420 has now been merged, this PR should be rebased so that it only includes the new changes/commits. Thanks!

MarieVerdonck committed:
Merge remote-tracking branch 'dspace-origin/master' into workflow-step-definitions

MarieVerdonck committed:
[DS-2646]: javadoc & SUBMIT_EDIT_METADATA to ProcessingAction

MarieVerdonck committed:
[DS-2646]: api/config/workflowdefinitions/defaultWorkflow/collections :
returns all non-mapped collections & all collections specifically mapped to defaultWorkflow
& Exception comment

MarieVerdonck committed:
[DS-2646]: checkstyle fix

KevinVdV committed:
Fixing configuration in the workflow-actions.xml test file

KevinVdV committed:
Merge branch 'master' into workflow-step-definitions

KevinVdV commented:
@benbosman Your feedback has been processed & @tdonohue the latest master has been merged into this branch as well, so it now only contains the relevant changes.

benbosman commented:
I've tested the functionality, and noticed:
* The endpoints are accessible for anonymous users. https://github.com/DSpace/Rest7Contract/blob/master/workflowdefinitions.md a.o. state: `All endpoints mentioned here require authentication, but no specific permissions`. This will need to be resolved
* I had to manually perform https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V7.0_2019.05.02__DS-4239-workflow-xml-migration.sql because my DB didn't contain the latest master yet
* Once the flyway migration was fixed, the workflow still works correctly

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/WorkflowDefinitionController.java:
```diff
@@ -0,0 +1,76 @@
+ */
+@RestController
+@RequestMapping("/api/" + WorkflowDefinitionRest.CATEGORY + "/" + WorkflowDefinitionRest.NAME_PLURAL)
+public class WorkflowDefinitionController {
```
This controller is misnamed & seems to be mapping the same request path as `WorkflowDefinitionRestRepository` (and `WorkflowDefinitionRest`).   I'd recommend renaming this to `WorkflowDefinitionCollectionController` (or similar), as this controller is specifically managing the Collection relationship here.  We can also then add a more specific `RequestMapping`...along the lines of:

```
@RequestMapping("/api/" + WorkflowDefinitionRest.CATEGORY + "/" + WorkflowDefinitionRest.NAME_PLURAL + "/{workflowName}/collections")
public class WorkflowDefinitionCollectionsController {
```

tdonohue replied:
Update: it's also worth considering whether this controller could be replaced with a `LinkRepository` to obtain the `collections` subresource.  That would better align this PR with the Projections work in #2625  (which has converted many similar Controller methods into LinkRepository classes)

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/test/AbstractWorkflowDefinitionIntegrationTest.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * @author Maria Verdonck (Atmire) on 17/12/2019
+ */
+public class AbstractWorkflowDefinitionIntegrationTest extends AbstractControllerIntegrationTest {
```
This class seems unnecessary to me, as it's just calling the `setup()` of `AbstractControllerIntegrationTest`.  Is there a reason why we don't just remove this class and use `AbstractControllerIntegrationTest` directly?

MarieVerdonck replied:
Removed, wasn't used

tdonohue reviewed (commented):
I've given this a code review today. Overall, it looks good. Just a few minor comments (see below).  I have not had a chance to test it out yet, but will do so.


UPDATE: I've tested this PR.  The endpoints all appear to work as described.  However, as noted by @benbosman  *all endpoints* (`/api/config/workflowdefinitions`, `/api/config/workflowsteps` and `/api/config/workflowactions`) *are accessible anonymously*.  This is unexpected behavior.  

To correct this, I suspect all the various `*RestRepository` classes simply need their methods to be annotated with `@PreAuthorize("hasAuthority('AUTHENTICATED')")`.  That should allow all to only be accessible to anyone who can successfully authenticate.

MarieVerdonck committed:
[DS-2646]: Workflow Endpoints now need authenticated access + tests:
WorkflowDefinition, WorkflowActions & WorkflowSteps Endpoints

MarieVerdonck committed:
[DS-2646]: /workflowdefinitions/<:name>/collections via @LinkRest (Controller to LinkRepository)
and /steps set embedOptional to true
> /workflowdefinitions/<:name> has links of subresources, but no embeds

MarieVerdonck committed:
[DS-2646]: checkstyle

MarieVerdonck committed:
[DS-2646]: test fix

MarieVerdonck commented:
@benbosman & @tdonohue : 
- WorkflowDefinition, WorkflowActions & WorkflowSteps Endpoints now need authenticated access + tests added
- /workflowdefinitions/<:name>/collections via @LinkRest (so Controller to LinkRepository) and /steps set embedOptional to true
    - With WorkflowDefintionRestRepository#findOne using the ListProjection (via utils.obtainProjection(true)) all its "sub resources" (steps and collections) have a link, but are not embedded, as long as they have the embedOptional set to true (if set to false, the default, they are embedded)
    - ListProjection should maybe be renamed to FlatProjection or NoEmbedProjection since its only defining feature is the lack of embedded resources when the embedding is optional
    - This allowed the HalLinkFactory that added the collections link manually to be removed

tdonohue reviewed (commented):
@MarieVerdonck : Thanks for the updates!

I've retested today and noticed a few small issues that I must have overlooked before.

1.  The `/api/config/workflowdefinitions` endpoint seems to return a list of `workflowDefinitionResources`.  I expected this to be a list of `workflowdefinitions`.  This would more align with other endpoints (e.g. the `/api/core/collections` endpoint returns a list of `collections` and not `collectionResources`).
2. I'm also seeing this same sort of behavior (lists appended with `Resources`) when looking at other endpoints that return lists.  For example: `/api/config/workflowsteps/reviewstep/workflowactions` returns a list of `workflowActionResources` instead of an expected list of `workflowactions`.
3. If I go to `/api/config/workflowdefinitions/defaultWorkflow/steps`, I'm getting a `204 No Content` response, which implies the `defaultWorkflow` has no steps, even though it's configured to have 3 steps: https://github.com/DSpace/DSpace/blob/master/dspace/config/spring/api/workflow.xml#L21-L27

I'm still doing some testing, but I wanted to report what I've found so far.  Overall the code is looking good, though and all the endpoints are now access restricted.

KevinVdV commented:
@tdonohue For points 1 & 2 in your feedback, the fact that the `/api/config/workflowdefinitions` & `/api/config/workflowsteps/reviewstep/workflowactions` return resources instead of Rest objects is managed globally & is so for all Repositories. The wrappings are done here: 
* https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java#L231
* https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java#L903

We will be looking into the third point of feedback.

tdonohue commented:
@KevinVdV : I'm not sure I understand then why *only* these new endpoints (`workflowdefinitions` and `workflowactions`) are embedded as `Resources` while all other existing endpoints return Rest objects?  Looking at all other endpoints, I'm not seeing others that embed `Resources`.... The `/api/core/items` endpoint returns a list of `items`, the `/api/core/collections` endpoint returns a list of `collections`, the  `/api/core/metadatafields` returns a list of `metadatafields`, the `/api/config/submissionforms` returns a list of `submissionforms`.  So, why should`/api/config/workflowdefinitions` return a list of `workflowDefinitionResources`?  It seems like an inconsistency that only occurs within the new endpoints in this PR.

KevinVdV commented:
@tdonohue You are correct, you were talking about the rest output while I was thinking about the actual objects in the code. We will look into this.

MarieVerdonck committed:
[DS-2646]: /workflowdefintions/<:name>/steps fixed + new tests &
endpoint result names changed to without Resource

MarieVerdonck commented:
@tdonohue & @KevinVdV 
Apparently in RestResourceController it gets a sub resource from its embedded resources RestResourceController#findRelInternal L849 resource.getEmbeddedResources().get(rel) instead of resource.getContent.get... (here getSteps), so I’m back at linking without embedding is not possible unless the result of the link request is handled in a separate method like /collections is in getCollections of WorkflowDefinitionCollectionsLinkRepository… ? :/
> Added tests for /steps endpoint, so this doesn’t get overlooked again 

Endpoint result names were with the Resource camelcase ending, because in the Workflow…Resource class this wasn’t replaced by a @RelNameDSpaceResource annotation

MarieVerdonck committed:
[DS-2646]: test and checkstyle fixes

KevinVdV committed:
Merge remote-tracking branch 'upstream/master' into workflow-step-definitions

MarieVerdonck committed:
68860: test fixes after merging latest dspace-origin/master

KevinVdV commented:
@tdonohue We fixed your test feedback & merged this branch with the latest master, so this PR is ready for further review.

tdonohue commented:
@KevinVdV and @MarieVerdonck :  I've tested this again today, and while some of my initial feedback has been resolved, I'm still seeing some oddities in how the `workflowdefinitions` endpoint works.  It's not seeming to return the same content as the workflow.xml configuration: https://github.com/DSpace/DSpace/blob/master/dspace/config/spring/api/workflow.xml

For example:

1. `/api/config/workflowdefinitions/defaultWorkflow/steps` still strangely returns a 204 (No Content) response for me, even though the `defaultWorkflow` has 3 steps defined in the `workflow.xml`
2. While the `/api/config/workflowsteps/<:id>` endpoint seems to let me load any step by it's ID, all those steps seem to have empty actions.  For example, `http://localhost:8080/server/api/config/workflowsteps/editstep/workflowactions` returns an 204 (No Content) response, when it has one action defined in workflow.xml

I will say that the `/api/config/workflowactions/` endpoint seems to work correctly, although it has no subresources or `_links` to other endpoints.  The issue seems to be with the subresources / links in both  the `workflowdefinitions` and `workflowsteps`.

As a sidenote, this PR also needs another rebase from a minor merge conflict.

MarieVerdonck committed:
Merge remote-tracking branch 'dspace-origin/master' into workflow-step-definitions

# Conflicts:
#	dspace-api/src/test/java/org/dspace/xmlworkflow/XmlWorkflowFactoryTest.java

MarieVerdonck committed:
test fix after master merge

MarieVerdonck committed:
constant for expected http status

MarieVerdonck committed:
LinkRepositories so that subresources paths work without projection=full (for workflowdefinitions/defaultWorkflow/steps and /workflowsteps/editstep/actions)

MarieVerdonck committed:
missing licenses

MarieVerdonck committed:
checkstyle

MarieVerdonck commented:
@tdonohue 
This used to work that @LinkRest method was sufficient to make the path of a subresource like /steps work, but now it seems that only works if you add ?projection=full or create a separate method in a LinkRepository to which the @LinkRest can refer… 
Linkrepositories for workflowdefinitions/defaultWorkflow/steps and /workflowsteps/editstep/workflowactions are now added so this works without adding ?projection=full

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkflowDefinitionRestRepositoryIT.java:
```diff
@@ -239,7 +240,7 @@ public void getWorkflowDefinitionByCollectionId_nonValidUUID() throws Exception
         getClient(token).perform(get(WORKFLOW_DEFINITIONS_ENDPOINT + "/search/findByCollection?uuid=" + nonValidUUID))
             //We expect a 400 Illegal Argument Exception cannot convert UUID
-            .andExpect(status().is(400))
+            .andExpect(status().is(HttpStatus.SC_BAD_REQUEST))
```
@MarieVerdonck : This could be simplified by writing
`.andExpect(status().isBadRequest());`   You'll see similar code in other ITs in our codebase.

tdonohue reviewed (approved):
👍 I tested this and it now works as described. Thanks @MarieVerdonck .  That said, I have one last tiny change to suggest (from a recent commit) -- see inline below. 

Overall though, this code looks good and works.

benbosman reviewed (approved):
I've reviewed the code, all issues I've identified have been resolved and I couldn't find any new issues.
I've also tested the code, and it worked as expected based on the contract.

MarieVerdonck committed:
is(HttpStatus.SC_BAD_REQUEST) to isBadRequest()
